### DISCUSSION TITLE
Is it possible to add xxhash32 and xxhash64 functions?

### DISCUSSION THREAD

**Discussion by usatenko** [duckdb/duckdb#4443] [General]
I would need compatibility with Clickhouse where xxhash is widely used. Is it possible to have in a near future?

**Comment by lnkuiper** [comment]
Looking at the [plethora of hash functions that ClickHouse provides](https://clickhouse.com/docs/en/sql-reference/functions/hash-functions/), I don't think that all of these hash functions should end up in the main DuckDB source code.

I think that adding a `hash` extension that includes these hash functions would be a good solution. We can then simply copy/paste the source code of these hash functions into the extension, and add a scalar function for each of them.

**Comment by usatenko** [comment]
The problem with your approach is that your hash is a blackbox, I saw recently you just switched the algorithm behind it, for end users it means all the hashes they rely on should be regenerated, whereas if they use specific hash function they may be sure they get what they expect.

**Reply by lnkuiper** [reply]
The hash function is the hash function we use internally, so I would not consider that a problem at all. It is useful to expose the internal hash function as a scalar function for, e.g., debugging purposes.

The problem that you're having is that you want to use this hash function as if it is a transparent, reliable hash function that is compatible with other systems. It's not, and it should not be used in that way.

We can implement the feature that you're requesting, and I think the best way to do it is with an extension so that the DuckDB source tree does not get cluttered with all kinds of libraries for hash functions.

**Comment by usatenko** [comment]
I actually think some set of the hash functions explicitly named should be in the base source to prevent misunderstandings

**Comment by usatenko** [comment]
But, anyway I will be very thankful to just have them available to use
